exporter/alibabacloudlogserviceexporter/tracedata_to_logservice.go

```diff
@@ -15,393 +15,225 @@
-			contents = append(contents, &sls.LogContent{
-				Key:   proto.String(processTagsPrefix + opencensusCoreLibVersion),
-				Value: proto.String(ocLib.CoreLibraryVersion),
-			})
-		}
-	}
+func resourceSpansToLogServiceData(resourceSpans pdata.ResourceSpans) []*sls.Log {
+	resourceContents := resourceToLogContents(resourceSpans.Resource())
+	insLibSpansSlice := resourceSpans.InstrumentationLibrarySpans()
+	var slsLogs []*sls.Log
+	for i := 0; i < insLibSpansSlice.Len(); i++ {
+		insLibSpans := insLibSpansSlice.At(i)
+		if insLibSpans.IsNil() {
+			continue
+		}
+		instrumentationLibraryContents := instrumentationLibraryToLogContents(insLibSpans.InstrumentationLibrary())
+		spans := insLibSpans.Spans()
+		for j := 0; j < spans.Len(); j++ {
+			span := spans.At(j)
+			if span.IsNil() {
+				continue
+			}
+			if slsLog := spanToLogServiceData(span, resourceContents, instrumentationLibraryContents); slsLog != nil {
+				slsLogs = append(slsLogs, slsLog)
+			}
+		}
+	}
+	return slsLogs
+}
 
-	var serviceName string
-	if node.ServiceInfo != nil && node.ServiceInfo.Name != "" {
-		serviceName = node.ServiceInfo.Name
+func spanToLogServiceData(span pdata.Span, resourceContents, instrumentationLibraryContents []*sls.LogContent) *sls.Log {
+	timeNano := int64(span.EndTime())
+	if timeNano == 0 {
+		timeNano = time.Now().UnixNano()
 	}
-
-	if resource != nil {
-		resourceType := resource.GetType()
-		if resourceType != "" {
-			contents = append(contents, &sls.LogContent{
-				Key:   proto.String(processTagsPrefix + opencensusResourceType),
-				Value: proto.String(resourceType),
-			})
-		}
-		for k, v := range resource.GetLabels() {
-			contents = append(contents, &sls.LogContent{
-				Key:   proto.String(processTagsPrefix + k),
-				Value: proto.String(v),
-			})
-		}
+	slsLog := sls.Log{
+		Time: proto.Uint32(uint32(timeNano / 1000 / 1000 / 1000)),
 	}
+	// pre alloc, refine if logContent's len > 16
+	preAllocCount := 16
+	slsLog.Contents = make([]*sls.LogContent, 0, preAllocCount+len(resourceContents)+len(instrumentationLibraryContents))
+	contentsBuffer := make([]sls.LogContent, 0, preAllocCount)
 
-	if serviceName == "" && len(contents) == 0 {
-		// No info to put in the process...
-		return nil
-	}
+	slsLog.Contents = append(slsLog.Contents, resourceContents...)
+	slsLog.Contents = append(slsLog.Contents, instrumentationLibraryContents...)
 
-	contents = append(contents, &sls.LogContent{
-		Key:   proto.String(serviceNameField),
-		Value: proto.String(serviceName),
+	contentsBuffer = append(contentsBuffer, sls.LogContent{
+		Key:   proto.String(traceIDField),
+		Value: proto.String(span.TraceID().HexString()),
+	})
+	contentsBuffer = append(contentsBuffer, sls.LogContent{
+		Key:   proto.String(spanIDField),
+		Value: proto.String(span.SpanID().HexString()),
+	})
+	contentsBuffer = append(contentsBuffer, sls.LogContent{
+		Key:   proto.String(parentSpanIDField),
+		Value: proto.String(span.ParentSpanID().HexString()),
```

Comment:
ParentSpanID could be empty when it's a root span, should we check it?

Reply:
if ParentSpanID is not valid, the return "", it is compatible for log service

Reply:
i have added a line comment about this